packages/actor-init-query/lib/HttpServiceSparqlEndpoint.ts

```diff
@@ -1,73 +1,344 @@
+
+    // TODO: Split the media types into pools based on response.resultType and do content negotiation over the
+    // appropriate pool only, then remove this code below, because it is a workaround to force default media types
+    // unless the HTTP client explicitly requests a specific one. The client can still request an incompatible format,
+    // but for cases like "Accept: */*" the code below will pick a media type that will work.
+    if (negotiatedVariant && negotiatedVariant.qts > 2) {
+      return negotiatedVariant.type;
+    }
+
+    switch (result.resultType) {
+      case 'bindings':
+        return 'application/sparql-results+json';
+      case 'quads':
+        return 'application/n-quads';
+      default:
+        return 'simple';
+    }
+  }
+
+  /**
+   * Extracts the SPARQL protocol operation from an incoming HTTP request.
+   * @param {URL} url The parsed request URL.
+   * @param {IncomingMessage} request The incoming HTTP request.
+   * @returns {ISparqlOperation} The parsed SPARQL protocol operation.
+   */
+  public async parseOperation(url: URL, request: IncomingMessage): Promise<ISparqlOperation> {
+    switch (request.method) {
+      case 'GET':
+      case 'HEAD':
+      case 'OPTIONS':
+        if (url.searchParams.has('query')) {
+          return {
+            type: 'query',
+            queryString: url.searchParams.get('query')!,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        if (url.searchParams.has('update')) {
+          return {
+            type: 'update',
+            queryString: url.searchParams.get('update')!,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        return {
+          type: 'sd',
+          queryString: '',
+          context: this.parseOperationParams(url.searchParams),
+        };
+      case 'POST':
+        // eslint-disable-next-line no-case-declarations
+        const requestBody = await this.readRequestBody(request);
+        if (requestBody.contentType.includes('application/sparql-query')) {
+          return {
+            type: 'query',
+            queryString: requestBody.content,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        if (requestBody.contentType.includes('application/sparql-update')) {
+          return {
+            type: 'update',
+            queryString: requestBody.content,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        if (requestBody.contentType.includes('application/x-www-form-urlencoded')) {
+          const requestBodyParams = new URLSearchParams(requestBody.content);
+          let requestBodyContext: QueryStringContext | undefined;
+          if (requestBodyParams.has('context')) {
+            try {
+              requestBodyContext = JSON.parse(requestBodyParams.get('context')!);
+            } catch {
+              break;
```

Comment:
JSON parse error message is getting lost.

Reply:
True, it could be reported to the client as invalid request!